Source/Core/VideoBackends/Software/SWVertexLoader.cpp

```diff
@@ -212,6 +210,42 @@ static void ReadVertexAttribute(T* dst, DataReader src, const AttributeFormat& f
   }
 }
 
+static void ParseColorAttributes(InputVertexData* dst, DataReader& src,
+                                 const PortableVertexDeclaration& vdec)
+{
+  bool use_color_1 = vdec.colors[0].enable && vdec.colors[1].enable;
+  for (size_t i = 0; i < 2; i++)
```

Comment:
I feel like the for loop is not gaining you very much due to the special casing (and the fact you only have 2 elements).

Based on my understanding, I think this would be ideal:

```
auto set_defaults = [dst](InputVertexData* dst type, int i)[]
{
  // Default on the alpha channel seems to depend on the number of components in the vertex format
  dst->color[i][0] = g_main_cp_state.vtx_attr[g_main_cp_state.last_id].g0.ColorElements[i] == 0 ? 255 : 0;
  dst->color[i][1] = 255;
  dst->color[i][2] = 255;
  dst->color[i][3] = 255;
};

if (vdec.colors[0].enable)
{
  ReadVertexAttribute<u8>(dst->color[0], src, vdec.colors[0], 0, 4, true);

  if (vdec.colors[1].enable;)
  {
    ReadVertexAttribute<u8>(dst->color[1], src, vdec.colors[1], 0, 4, true);
  }
  else
  {
    // Default channel 1
    set_defaults(dst, 1);
  }
}
// Only channel 1 is enabled
else if (vdec.colors[1].enable;)
{
  ReadVertexAttribute<u8>(dst->color[i], src, vdec.colors[1], 0, 4, true);
}
else
{
  // Set defaults for both channels
  set_defaults(dst, 0);
  set_defaults(dst, 1);
}
```

If we can't or don't want to use a lambda due to needing to change the "ColorElements" into an array, I'd say just copy and paste the 4 lines.  I know copy+paste is generally bad but I think it gives clearer meaning.